Simulation concepts doc (#4808)
cosmos/cosmos-sdk: docs/concepts/using-the-sdk/simulation.md

The Cosmos SDK offers a full fledged simulation framework to fuzz test every message defined by a module.

This functionality is provided by the[`SimApp`](https://github.com/cosmos/cosmos-sdk/blob/master/simapp/app.go),
which is a dummy application that is used for running the [`simulation`](https://github.com/cosmos/cosmos-sdk/tree/master/x/simulation) module.
This module defines all the simulation logic as well as the operations for randomized parameters like accounts, balances etc.

## Goals

The blockchain simulator tests how the blockchain application would behave under real life circumstances by generating and sending randomized messages.
The goal of this is to detect and debug failures that could halt a live chain, by providing logs and statistics about the operations run by the simulator as well as exporting the latest application state when a failure was found.

Its main difference with integration testing is that the simulator app allows you to pass parameters to customize the chain that's being simulated.
This comes in handy when trying to reproduce bugs that were generated in the provided operations (randomized or not).

## Simulation commands

> The simulation app has different commands, each of it tests a different failure type:

Reviewer: "The simulation app has different commands, each of `which` tests a different failure type:

> The simulation app has different commands, each of it tests a different failure type:

Reviewer 2: ```suggestion
The simulation app has different commands, each of which tests a different failure type:
```

Check the full list of flags [here](https://github.com/cosmos/cosmos-sdk/blob/adf6ddd4a807c8363e33083a3281f6a5e112ab89/simapp/sim_test.go#L34-L50).

## Simulator Modes

In addition to the various inputs and commands, the simulator runs in three modes:

1. Completely random where the initial state, module parameters and simulation parameters are **pseudo-randomly generated**.
2. From a `genesis.json` file where the initial state and the module parameters are defined.
This mode is helpful for running simulations on a known state such as a live network export where a new (mostly likely breaking) version of the application needs to be tested.
3. From a `params.json` file where the initial state is pseudo-randomly generated but the module and simulation parameters can be provided manually.
This allows for a more controlled and deterministic simulation setup while allowing the state space to still be pseudo-randomly simulated. The list of available parameters is listed [here](https://github.com/cosmos/cosmos-sdk/blob/adf6ddd4a807c8363e33083a3281f6a5e112ab89/x/simulation/params.go#L170-L178).

::: tip

> These modes are not excluding between each other. So you can for example run a randomly generated genesis state (`1`) with manually generated simulation params (`3`).

Reviewer: "These modes are not `mutually exclusive`"

> These modes are not excluding between each other. So you can for example run a randomly generated genesis state (`1`) with manually generated simulation params (`3`).

Reviewer 2: ```suggestion
These modes are not mutually exclusive. So you can for example run a randomly generated genesis state (`1`) with manually generated simulation params (`3`).
```
